File: javascript/binary-search-tree/binary-search-tree.js
export class BinarySearchTree {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }

  insert(data) {
    if (data <= this.data) {
      if (!this.left) {
        this.left = new BinarySearchTree(data);
      } else {
        this.left.insert(data);
      }
    } else if (!this.right) {
      this.right = new BinarySearchTree(data);
    } else {
      this.right.insert(data);
    }
  }

  each(cb, root = this) {
    if (root) {
      this.each(cb, root.left);
      cb(root.data);
      this.each(cb, root.right);
    }
  }
}

File: dart/binary-search-tree/lib/binary_search_tree.dart
class TreeNode {
  String data;
  TreeNode left;
  TreeNode right;

  TreeNode(String data) {
    this.data = data;
  }
}

class BinarySearchTree {
  TreeNode root;
  List<String> sortedTree;

  BinarySearchTree(String input) {
    insert(input);
  }

  void insert(String data, [TreeNode root = null]) {
    // is first call to insert (at root)?
    if (root == null) {
      root = this.root;
    }

    // first element in tree?
    if (root == null) {
      this.root = TreeNode(data);
      return;
    }

    // find correct location, recursively traversing tree
    if (data.compareTo(root.data) <= 0) {
      if (root.left == null) {
        root.left = TreeNode(data);
      } else {
        insert(data, root.left);
      }
    } else {
      if (root.right == null) {
        root.right = TreeNode(data);
      } else {
        insert(data, root.right);
      }
    }
  }

  List<String> get sortedData {
    sortedTree = [];
    inOrder(this.root);
    return sortedTree;
  }

  void inOrder(TreeNode root) {
    if (root == null) {
      return;
    }
    if (root.left != null) {
      inOrder(root.left);
    }
    sortedTree.add(root.data);
    if (root.right != null) {
      inOrder(root.right);
    }
  }
}
